### PR TITLE
Error on duplicate test output dirs

### LINKED ISSUE #16490: Ensure tests write to separate output directories
See discussion on https://github.com/dotnet/sdk/pull/16425

### PR DESCRIPTION
**Pull request by sfoslund**
Fixes https://github.com/dotnet/sdk/issues/16490

### CODE REVIEW

File: src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs [hunk 1]
```diff
@@ -128,14 +130,20 @@ public string GetAndValidateTestProjectDirectory(string testProjectName, string
                 using (var sha256 = SHA256.Create())
                 {
                     var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(directoryName.ToString()));
+                    var rand = Guid.NewGuid().ToString("N").Substring(0, 5);
```
Comment (joeloff) [inline src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:133]: Path.GetRandomFileName perhaps, or do you need a shorter random string?

**Review by dsplaisted — COMMENTED** [review]
I would prefer not to use random paths.  If we have to, I'd prefer to do it only in CI runs so that you have a predictable path locally, and so that you don't have to keep cleaning up your test folder to stop it from growing without bounds.

Alternatively, could we have a mode that makes it an error if the test folder already exists when we try to create it?  Then that would help us find all of the tests that have overlapping test folders.

**Comment by wli3** [comment]
Per https://github.com/dotnet/sdk/pull/16425#issuecomment-805219482. We've been trying to get a stable path without conflict for years but we still cannot get it right. And I think it is good enough that the test can print out the temp directory,

**Comment by sfoslund** [comment]
> We've been trying to get a stable path without conflict for years but we still cannot get it right. And I think it is good enough that the test can print out the temp directory

So you'd prefer we make it random and print out the directory name? Or print the name that's already being generated?

**Comment by dsplaisted** [comment]
> Per [#16425 (comment)](https://github.com/dotnet/sdk/pull/16425#issuecomment-805219482). We've been trying to get a stable path without conflict for years but we still cannot get it right. And I think it is good enough that the test can print out the temp directory,

We've just been playing whack-a-mole, when we have a test failing that looks like it might be due to this we look for another test that would use the same folder.  We should be able to deterministically find all the conflicts by adding logic to error out if the test folder already exists when we try to create it, and running that with an empty test folder.

**Comment by danmoseley** [comment]
Note there is a middle path, which is deterministic + a random suffix. Eg., c:\foo\bar......\mytestmethod.ASD2435SDF ... that avoids collisions, but still leaves a hint as to which test.

I have no preference.

**Comment by sfoslund** [comment]
> We should be able to deterministically find all the conflicts by adding logic to error out if the test folder already exists when we try to create it, and running that with an empty test folder.

I did a test run with this error and there are quite a lot of individual cases we'll have to handle with this approach. Do we want to allow theories to use the same test folders?

**Comment by dsplaisted** [comment]
> > We should be able to deterministically find all the conflicts by adding logic to error out if the test folder already exists when we try to create it, and running that with an empty test folder.
> 
> I did a test run with this error and there are quite a lot of individual cases we'll have to handle with this approach. Do we want to allow theories to use the same test folders?

Yes, ideally each instance of a theory would use a separate test folder.  The `identifier` parameter to `CreateTestProject` or similar methods can be used for this.

It's possible that a theory not using different test folders for each instance is less likely to cause flaky failures than an entirely separate test case, as I believe unrelated test cases can be run concurrently while instances of theories are run sequentially.

File: src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs [hunk 2]
```diff
@@ -139,7 +139,14 @@ public string GetAndValidateTestProjectDirectory(string testProjectName, string
                 }
             }
 
-            return Path.Combine(baseDirectory, directoryName.ToString());
+            var directoryPath = Path.Combine(baseDirectory, directoryName.ToString());
+
+            if (Directory.Exists(directoryPath))
+            {
+                throw new Exception($"Test dir {directoryPath} already exists");
+            }
```
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:147]: By default, we should only error out when running in CI.  Otherwise rerunning tests locally will fail by default unless you delete the test directory manually each time.

**Review by dsplaisted — COMMENTED** [review]
Thanks a bunch for fixing all of these!

File: src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs (hunk 2 above)
Comment (sfoslund) [reply src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:147]: Do we have a flag for if the tests are running in CI already?

Comment (dsplaisted) [reply src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:147]: I think there is an MSBuild property or environment variable.  Try downloading a binlog from a CI run.  Also when the tests are run in helix there may be something that tells us about that.  @wli3

Comment (wli3) [reply src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:147]: I think we can make it an error on default. And add an extra "set environment" and a new env in .\artifacts\sdk-build-env.bat and .\artifacts\sdk-build-env.sh

Comment (wli3) [reply src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:147]: we need to run .\artifacts\sdk-build-env.bat to build sdk anyway

Comment (dsplaisted) [reply src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:147]: > we need to run .\artifacts\sdk-build-env.bat to build sdk anyway

I basically always run it but a lot of the time it's not necessary.  I imagine most people that try to use our repo will just open the .sln in Visual Studio and they'll want the tests to work.

Comment (sfoslund) [reply src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:147]: Agreed, I'm went ahead and conditioned it based on the `ContinuousIntegrationBuild` property.

Comment (wli3) [reply src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs:147]: >  will just open the .sln in Visual Studio and they'll want the tests to work.

nothing will work by open .sln in visual studio

**Comment by sfoslund** [comment]
@dsplaisted @wli3 is this good to go?

**Review by dsplaisted — APPROVED** [review]
